rewrite-java-17/src/main/java/org/openrewrite/java/Java17Parser.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2020 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java;
+
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.java.internal.JavaTypeCache;
+import org.openrewrite.java.marker.JavaSourceSet;
+import org.openrewrite.java.tree.J;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.Charset;
+import java.nio.file.*;
+import java.util.*;
+
+public class Java17Parser implements JavaParser {
+    private final JavaParser delegate;
+
+    Java17Parser(JavaParser delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public List<J.CompilationUnit> parseInputs(Iterable<Input> sourceFiles, @Nullable Path relativeTo, ExecutionContext ctx) {
+        return delegate.parseInputs(sourceFiles, relativeTo, ctx);
+    }
+
+    @Override
+    public JavaParser reset() {
+        return delegate.reset();
+    }
+
+    @Override
+    public void setClasspath(Collection<Path> classpath) {
+        delegate.setClasspath(classpath);
+    }
+
+    @Override
+    public void setSourceSet(String sourceSet) {
+        delegate.setSourceSet(sourceSet);
+    }
+
+    @Override
+    public JavaSourceSet getSourceSet(ExecutionContext ctx) {
+        return delegate.getSourceSet(ctx);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder extends JavaParser.Builder<Java17Parser, Builder> {
+
+        @Nullable
+        private static ClassLoader moduleClassLoader;
+
+        static synchronized void lazyInitClassLoaders() {
+            if (moduleClassLoader != null) {
+                return;
+            }
+
+            ClassLoader appClassLoader = Java17Parser.class.getClassLoader();
+            moduleClassLoader = new ModuleClassLoader(appClassLoader);
+
+        }
+
+        @Override
+        public Java17Parser build() {
+            lazyInitClassLoaders();
+
+            try {
+                // need to reverse this parent/child relationship
+                Class<?> parserImplementation = Class.forName("org.openrewrite.java.isolated.IsolatedJava17Parser", true, moduleClassLoader);
```

Comment:
Maybe follow the same naming as the reflection loaded Java8 parser? There we called it `ReloadableJava8Parser` and I think it was in the `.internal` package?

Reply:
I renamed the classes by prefixing them with "Reloadable". I left the classes in `org.openrewrite.java.isolated` as `rewrite-java` has classes in `org.openrewrite.java.internal` and we do not want those loaded by the new class loader. 

Right now the class loader uses the package name to determine which classes are loaded vs loaded by the parent. This was easier than enumerating specific classes (and nested classes)